regtest/README.md

```diff
@@ -0,0 +1,103 @@
+## Regtest Mode
+WARNING Experimental!
+The CLI can work in regtest mode, by locally running a `zcashd` and `lightwalletd`.
+This is now working with a simple zingo-cli invocation flag, with a little prior setup.
+
+There are pre-made directories in this repo to support ready use of regtest mode. These are found in the `/regtest/` subdirectory.
+
+Usage example:
+Copy your compiled `zcashd` `zcash-cli` and `lightwalletd` binaries to `zingolib/regtest/bin/` or set up symlinks, etc.
+
+There are default config files for these binaries already in place in `/zingolib/regtest/conf/` - which can be edited also.
+
+From your `zingolib/` directory, with, for example, a release build (`cargo build --release`), you can run:
+`./target/release/zingo-cli --regtest --data-dir regtest/datadir/zingo --server=127.0.0.1:9067`
```

Comment:
```
└─[$] <git:(regtest_mode*)> ./target/release/zingo-cli --regtest --data-dir regtest/datadir/zingo --server=127.0.0.1:9067
--datadir=/Users/skyl/Code/github/free2z/zingolabs/zingolib/regtest/datadir/zcash/
--conf=/Users/skyl/Code/github/free2z/zingolabs/zingolib/regtest/conf/zcash.conf
zcashd is starting in regtest mode, please standby about 10 seconds...
zcashd start section completed, zcashd should be running.
Standby, lightwalletd is about to start. This should only take a moment.
Using config file: /Users/skyl/Code/github/free2z/zingolabs/zingolib/regtest/conf/lightwalletdconf.yml
Starting insecure server
lwd start section completed, lightwalletd should be running!
Standby, Zingo-cli should be running in regtest mode momentarily...
Error during startup:Error with response: Status { code: Unavailable, message: "error trying to connect: tcp connect error: Connection refused (os error 61)", source: Some(hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 61, kind: ConnectionRefused, message: "Connection refused" }))) }
If you repeatedly run into this issue, you might have to restore your wallet from your seed phrase.
```

Reply:
Run again:

```
└─[$] <git:(regtest_mode*)> ./target/release/zingo-cli --regtest --data-dir regtest/datadir/zingo --server=127.0.0.1:9067
--datadir=/Users/skyl/Code/github/free2z/zingolabs/zingolib/regtest/datadir/zcash/
--conf=/Users/skyl/Code/github/free2z/zingolabs/zingolib/regtest/conf/zcash.conf
zcashd is starting in regtest mode, please standby about 10 seconds...
zcashd start section completed, zcashd should be running.
Standby, lightwalletd is about to start. This should only take a moment.
Using config file: /Users/skyl/Code/github/free2z/zingolabs/zingolib/regtest/conf/lightwalletdconf.yml
Starting insecure server
Lightwalletd died with a Fatal error. Check logfile for details.
lwd start section completed, lightwalletd should be running!
Standby, Zingo-cli should be running in regtest mode momentarily...
regtest detected and network set correctly!
Creating a new wallet
Lightclient connecting to http://127.0.0.1:9067/
Error with response: Status { code: Unknown, message: "cache is empty, server is probably not yet ready", metadata: MetadataMap { headers: {"content-type": "application/grpc"} }, source: None }
Ready!
(regtest) Block:0 (type 'help') >> 
```

Reply:
Run 3

```
└─[$] <git:(regtest_mode*)> ./target/release/zingo-cli --regtest --data-dir regtest/datadir/zingo --server=127.0.0.1:9067
--datadir=/Users/skyl/Code/github/free2z/zingolabs/zingolib/regtest/datadir/zcash/
--conf=/Users/skyl/Code/github/free2z/zingolabs/zingolib/regtest/conf/zcash.conf
zcashd is starting in regtest mode, please standby about 10 seconds...
zcashd start section completed, zcashd should be running.
Standby, lightwalletd is about to start. This should only take a moment.
Using config file: /Users/skyl/Code/github/free2z/zingolabs/zingolib/regtest/conf/lightwalletdconf.yml
Starting insecure server
Lightwalletd died with a Fatal error. Check logfile for details.
lwd start section completed, lightwalletd should be running!
Standby, Zingo-cli should be running in regtest mode momentarily...
regtest detected and network set correctly!
Lightclient connecting to http://127.0.0.1:9067/
Error with response: Status { code: Unknown, message: "cache is empty, server is probably not yet ready", metadata: MetadataMap { headers: {"content-type": "application/grpc"} }, source: None }
Ready!
```

Reply:
```
⇒  git status                   
On branch regtest_mode
Untracked files:
  (use "git add <file>..." to include in what will be committed)

        regtest/bin/lightwalletd
        regtest/bin/server.log
        regtest/bin/zcash-cli
        regtest/bin/zcashd
        regtest/datadir/lightwalletd/db/
        regtest/datadir/zcash/regtest/
        regtest/datadir/zingo/zingo-wallet.dat
        regtest/datadir/zingo/zingo-wallet.debug.log
        regtest/logs/lwd.log
```

Reply:
`Lightwalletd died with a Fatal error. Check logfile for details.` -> lwd.log seems to stream normal activity actually ...

seems to be working actually, despite the "Fatal error" message

```
└─[$] <git:(regtest_mode*)> ./target/release/zingo-cli --regtest --data-dir regtest/datadir/zingo --server=127.0.0.1:9067
--datadir=/Users/skyl/Code/github/free2z/zingolabs/zingolib/regtest/datadir/zcash/
--conf=/Users/skyl/Code/github/free2z/zingolabs/zingolib/regtest/conf/zcash.conf
zcashd is starting in regtest mode, please standby about 10 seconds...
zcashd start section completed, zcashd should be running.
Standby, lightwalletd is about to start. This should only take a moment.
Using config file: /Users/skyl/Code/github/free2z/zingolabs/zingolib/regtest/conf/lightwalletdconf.yml
Starting insecure server
Lightwalletd died with a Fatal error. Check logfile for details.
lwd start section completed, lightwalletd should be running!
Standby, Zingo-cli should be running in regtest mode momentarily...
regtest detected and network set correctly!
Lightclient connecting to http://127.0.0.1:9067/
Error with response: Status { code: Unknown, message: "cache is empty, server is probably not yet ready", metadata: MetadataMap { headers: {"content-type": "application/grpc"} }, source: None }
Ready!
(regtest) Block:0 (type 'help') >> help
Available commands:
new - Create a new address in this wallet
list - List all transactions in the wallet
seed - Display the seed phrase
encryptmessage - Encrypt a memo to be sent to a z-address offline
getoption - Get a wallet option
addresses - List all addresses in the wallet
decryptmessage - Attempt to decrypt a message with all the view keys in the wallet.
setoption - Set a wallet option
balance - Show the current ZEC balance in the wallet
send - Send ZEC to the given address
quit - Quit the lightwallet, saving state to disk
help - Lists all available commands
height - Get the latest block height that the wallet is at
info - Get the lightwalletd server's info
encrypt - Encrypt the wallet with a password
defaultfee - Returns the default fee in zats for outgoing transactions
clear - Clear the wallet state, rolling back the wallet to an empty state.
unlock - Unlock wallet encryption for spending
export - Export private key for wallet addresses
zecprice - Get the latest ZEC price in the wallet's currency (USD)
sendprogress - Get the progress of any send transactions that are currently computing
decrypt - Completely remove wallet encryption
import - Import spending or viewing keys into the wallet
lock - Lock a wallet that's been temporarily unlocked
rescan - Rescan the wallet, downloading and scanning all blocks and transactions
encryptionstatus - Check if the wallet is encrypted and if it is locked
sync - Download CompactBlocks and sync to the server
syncstatus - Get the sync status of the wallet
lasttxid - Show the latest TxId in the wallet
shield - Shield your transparent ZEC into a sapling address
notes - List all sapling notes and utxos in the wallet
save - Save wallet file to disk
(regtest) Block:0 (type 'help') >> addresses
{
  "z_addresses": [
    "zregtestsapling1dh4anz7h6kda7ggvm977c56vaycg84n0va3n07nm56ynysjzlaqzlx6w2j5gk6s5k2pv7ygl9q9"
  ],
  "t_addresses": [
    "tmSXYaU55ANrRUxDjbpDWjkT2akZ5jRgxdd"
  ],
  "o_addresses": []
}
(regtest) Block:0 (type 'help') >> nnotes
Unknown command : nnotes. Type 'help' for a list of commands
(regtest) Block:0 (type 'help') >> notes
{
  "unspent_notes": [],
  "pending_notes": [],
  "utxos": [],
  "pending_utxos": []
}
(regtest) Block:0 (type 'help') >> zecprice
Error: No price
(regtest) Block:0 (type 'help') >> seed
{
  "seed": "video payment ... hard",
  "birthday": 1
}
(regtest) Block:0 (type 'help') >> list
[]
(regtest) Block:0 (type 'help') >> new
No address type specified
Create a new address in this wallet
Usage:
new [z | t | o]

Example:
To create a new z address:
new z
(regtest) Block:0 (type 'help') >> 
```

Reply:
```
(regtest) Block:0 (type 'help') >> new o
[
  "uregtest1rp5974yuhsw3hq0fqw2ftkg7cmky3n9wu828065xn74zcrfudzjc0ykmgkh9xe83xtl6uphxh7pfs797ndaezrcaxfd74pk6hgy27x6e"
]
```

Reply:
I think you may have daemons still running after running the regtest mode, which might explain the sequence of errors of subsequent runs?

Currently those do not shut down when exiting zingo-cli, so you have to do that manually.

I have noticed that sometimes I had to run this cli more than once - maybe a db or something has to come up first, or takes more time initializing. I'll leave this convo open though your `tcp connect error: Connection refused (os error 61)` I don't think I've seen, this may have to do with local environment or containers?

Reply:
Circling back to say I saw `tcp connect error: Connection refused` when I was running more than one `lightwalletd` today, IIUC.